File: gems/ol-asciidoc/lib/ol-asciidoc.rb
require 'asciidoctor'
require 'asciidoctor/extensions'

class FeatureLinkMacro < Asciidoctor::Extensions::InlineMacroProcessor
  use_dsl
  named :feature

  def process parent, target, attrs

    (create_anchor parent, target, type: :link, target: "/docs/ref/feature/#" + target + ".html").convert
  end
end

class ConfigLinkMacro < Asciidoctor::Extensions::InlineMacroProcessor
  use_dsl
  named :config

  def process parent, target, attrs
    (create_anchor parent, target, type: :link, target: "/docs/ref/config/#rwlp_config_" + target + ".html").convert
  end
end

class JavadocLinkMacro < Asciidoctor::Extensions::InlineMacroProcessor
  use_dsl
  named :javadoc

  def process parent, target, attrs
    target
  end
end

Asciidoctor::Extensions.register do
  inline_macro FeatureLinkMacro
  inline_macro ConfigLinkMacro
  inline_macro JavadocLinkMacro
end

File: scripts/ol-asciidoc.js
const { posix: path } = require('path');

function DocLinkMacros (context, docType) {
  return function () {
    this.process((parent, target, attrs) => {
      const text = attrs['display'] || target;
      const pageId = path.join(path.dirname(context.file.src.relative), target);
      // NOTE the value of the path attribute is never used, so we can fake it
      const attributes = Opal.hash2(['refid', 'path'], { refid: pageId, path: pageId });
      if (docType == "javadoc") {
        return target;
      }
      else {
        return this.createInline(parent, 'anchor', text, { type: 'link', target: "/docs/latest/reference/" + docType + "/" + target + ".html", attributes });
      }
    })
  }
}

function register (registry, context) {
  registry.inlineMacro('config', DocLinkMacros(context, "config"));
  registry.inlineMacro('feature', DocLinkMacros(context, "feature"));
  registry.inlineMacro('javadoc', DocLinkMacros(context, "javadoc"));
}

module.exports.register = register;